src/plugins/data/public/search/errors/utils.ts

```diff
@@ -15,6 +15,15 @@ export function getFailedShards(err: KibanaServerError<any>): FailedShard | unde
   return failedShards ? failedShards[0] : undefined;
 }
 
+function getNestedCause(err: KibanaServerError<any>): Reason {
+  const { type, reason, caused_by: causedBy } = err.attributes || err;
```

Comment:
We should avoid introducing additional `any`s. Here is a possible alternative:

```ts
import type { ErrorCause } from '@elastic/elasticsearch/api/types';
...
function getNestedCause(err: KibanaServerError | ErrorCause): Reason {
  const attr = ((err as KibanaServerError).attributes || err) as ErrorCause;
  const { type, reason, caused_by: causedBy } = attr;
  if (causedBy) {
    return getNestedCause(causedBy);
  }
  return { type, reason };
}
```

We could improve `KibanaServerError` by defaulting the generic to `ErrorCause`. I'll open up an issue so that we can discuss that change separately. We could also add a new `isKibanaServerError` predicate to avoid using `as` as an improvement.

Reply:
Makes sense. There's some more type work to do, in Lens we did inspect quite a bit the EsError shape (for our use cases) but the type error source is not precise itself (I've discussed with @delvedor before about mapping all possible EsError in types but it's a big task to tackle right now).

It's definitely possible to patch this now and open a discussion about error types improvement later on.